### PR TITLE
Follow/unfollow thread backend

### PR DESCRIPTION
**Pull request by megrogan**

### CODE REVIEW

File: backend/canisters/community/impl/src/updates/follow_thread.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    let user_id = match state.data.members.get(caller) {
+        Some(member) if member.suspended.value => return UserSuspended,
+        Some(member) => member.user_id,
+        None => return UserNotInCommunity,
+    };
+
+    if let Some(channel) = state.data.channels.get_mut(&args.channel_id) {
+        match channel.chat.follow_thread(user_id, args.thread_root_message_index, now) {
+            FollowThreadResult::Success => {
+                handle_activity_notification(state);
```
Comment (hpeebles) [inline backend/canisters/community/impl/src/updates/follow_thread.rs:32]: This shouldn't mark the whole community as active. We should should `c2c_mark_community_updated_for_user`

File: backend/canisters/community/impl/src/updates/unfollow_thread.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    let user_id = match state.data.members.get(caller) {
+        Some(member) if member.suspended.value => return UserSuspended,
+        Some(member) => member.user_id,
+        None => return UserNotInCommunity,
+    };
+
+    if let Some(channel) = state.data.channels.get_mut(&args.channel_id) {
+        match channel.chat.unfollow_thread(user_id, args.thread_root_message_index, now) {
+            UnfollowThreadResult::Success => {
+                handle_activity_notification(state);
```
Comment (hpeebles) [inline backend/canisters/community/impl/src/updates/unfollow_thread.rs:32]: Same here.

File: backend/canisters/group/api/can.did [hunk 3]
```diff
@@ -563,6 +563,34 @@ type ResetInviteCodeResponse = variant {
     ChatFrozen;
 };
 
+type FollowThreadArgs = record {
```
Comment (hpeebles) [inline backend/canisters/group/api/can.did:566]: Maybe just call this `FollowUnfollowThreadArgs`?

File: backend/canisters/group/impl/src/updates/follow_thread.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    let user_id = match state.data.lookup_user_id(caller) {
+        Some(uid) => uid,
+        None => return UserNotInGroup,
+    };
+
+    let now = state.env.now();
+
+    match state.data.chat.follow_thread(user_id, args.thread_root_message_index, now) {
+        FollowThreadResult::Success => {
+            handle_activity_notification(state);
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/updates/follow_thread.rs:31]: `c2c_mark_group_updated_for_user`

File: backend/canisters/group/impl/src/updates/unfollow_thread.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    let user_id = match state.data.lookup_user_id(caller) {
+        Some(uid) => uid,
+        None => return UserNotInGroup,
+    };
+
+    let now = state.env.now();
+
+    match state.data.chat.unfollow_thread(user_id, args.thread_root_message_index, now) {
+        UnfollowThreadResult::Success => {
+            handle_activity_notification(state);
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/updates/unfollow_thread.rs]: `c2c_mark_group_updated_for_user`

File: backend/canisters/group/api/can.did (hunk 3 above)
Comment (megrogan) [reply backend/canisters/group/api/can.did:566]: I think this is fine and what we have done in general

File: backend/canisters/community/impl/src/updates/follow_thread.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    let user_id = match state.data.members.get(caller) {
+        Some(member) if member.suspended.value => return UserSuspended,
+        Some(member) => member.user_id,
+        None => return UserNotInCommunity,
+    };
+
+    if let Some(channel) = state.data.channels.get_mut(&args.channel_id) {
+        match channel.chat.follow_thread(user_id, args.thread_root_message_index, now) {
+            FollowThreadResult::Success => {
+                state.data.fire_and_forget_handler.send(
```
Comment (hpeebles) [inline backend/canisters/community/impl/src/updates/follow_thread.rs:34]: Maybe move this into `lib`? Since it is now used in multiple places.

**Review by hpeebles — APPROVED** [review]

File: backend/canisters/group/impl/src/lib.rs [hunk 7]
```diff
@@ -448,6 +449,14 @@ impl Data {
             .record(function_id, instructions_count, wasm_version, now);
     }
 
+    pub fn mark_group_updated_in_user_canister(&self, user_id: UserId) {
```
Comment (hpeebles) [inline backend/canisters/group/impl/src/lib.rs:452]: This is also used when toggling notifications (+ for communities)

**Pull request merged**